openlane/flows/synth_explore.py

```diff
@@ -0,0 +1,143 @@
+# limitations under the License.
+from __future__ import annotations
+from decimal import Decimal
+import os
+
+import rich
+import rich.table
+from typing import Dict, List, Optional, Tuple
+from concurrent.futures import Future
+
+from .flow import Flow
+from ..state import State
+from ..config import Config
+from ..logging import success
+from ..logging import options, console
+from ..steps import Step, Yosys, OpenROAD, StepError
+
+
+# "Synthesis Exploration" is a non-seqeuential flow that tries all synthesis
+# strategies and shows which ones yield the best area XOR delay
+@Flow.factory.register()
+class SynthesisExploration(Flow):
+    Steps = [
+        Yosys.Synthesis,
+        OpenROAD.CheckSDCFiles,
+        OpenROAD.STAPrePNR,
+    ]
+
+    def run(
+        self,
+        initial_state: State,
+        **kwargs,
+    ) -> Tuple[State, List[Step]]:
+        step_list: List[Step] = []
+
+        self.progress_bar.set_max_stage_count(1)
+
+        synth_futures: List[Tuple[Config, Future[State]]] = []
+        self.progress_bar.start_stage("Synthesis Exploration")
+
+        options.set_condensed_mode(True)
+
+        for strategy in [
+            "AREA 0",
+            "AREA 1",
+            "AREA 2",
+            "AREA 3",
+            "DELAY 0",
+            "DELAY 1",
+            "DELAY 2",
+            "DELAY 3",
+            "DELAY 4",
+        ]:
+            config = self.config.copy(SYNTH_STRATEGY=strategy)
+
+            synth_step = Yosys.Synthesis(
+                config,
+                id=f"synthesis-{strategy}",
+                state_in=initial_state,
+            )
+            synth_future = self.start_step_async(synth_step)
+            step_list.append(synth_step)
+
+            sdc_step = OpenROAD.CheckSDCFiles(
+                config,
+                id=f"sdc-{strategy}",
+                state_in=synth_future,
+            )
+            sdc_future = self.start_step_async(sdc_step)
+            step_list.append(sdc_step)
+
+            sta_step = OpenROAD.STAPrePNR(
+                config,
+                state_in=sdc_future,
+                id=f"sta-{strategy}",
+            )
+
+            step_list.append(sta_step)
+            sta_future = self.start_step_async(sta_step)
+
+            synth_futures.append((config, sta_future))
+
+        results: Dict[str, Optional[Tuple[Decimal, Decimal, Decimal, Decimal]]] = {}
+        for config, future in synth_futures:
+            strategy = config["SYNTH_STRATEGY"]
+            results[strategy] = None
+            try:
+                state = future.result()
+                results[strategy] = (
+                    state.metrics["design__instance__count"],
+                    state.metrics["design__instance__area"],
+                    state.metrics["timing__setup__ws"],
+                    state.metrics["timing__setup__tns"],
+                )
+            except StepError:
+                pass  # None == failure
```

Comment:
raise an error?

Reply:
Hmm then maybe the table should indicate it somehow?

Reply:
It does, with the word "Failed"